<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title> Introduction |  Cluster Flow</title>
	<meta name="description" content="A pipelining tool to automate and standardise bioinformatics analyses on cluster environments"/>
	<meta name="author" content="Phil Ewels"/>
	<link rel="shortcut icon" href="/favicon.ico">

	<!-- Bootstrap and Google Fonts -->
	<link href="/resources/bootstrap/css/bootstrap.min.css" rel="stylesheet">
	<link href='http://fonts.googleapis.com/css?family=Lato:400,700,400italic' rel='stylesheet' type='text/css'>
	<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
	<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
	<!--[if lt IE 9]>
		<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
		<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
	<![endif]-->

	<!-- Site Stylesheet -->
	<link rel="stylesheet" href="/style.css">
</head>
<body>

<header>
	<section class="logo">
		<h1><a href="/"><img src="/resources/img/CF_logo.png" title="Cluster Flow"></a></h1>
		<p>A pipelining tool to automate and standardise bioinformatics analyses on cluster environments.</p>
	</section>
	<section class="download">
		<p>
			<a class="btn btn-default" href="/" title="Download Cluster Flow"><i class="glyphicon glyphicon-folder-open"></i> &nbsp; &nbsp; Download Cluster Flow</a>
			<a class="text-link" href="/" title="View on GitHub"><img src="/resources/img/GitHub-Mark/PNG/GitHub-Mark-Light-16px.png" title="View on GitHub"> &nbsp; View on GitHub</a>
		</p>
	</section>
	<nav id="nav">
		<ul>
			<li><a href="/introduction/">Introduction</a></li>
			<li><a href="/installation/">Installation Instructions</a></li>
			<li><a href="/usage/">General Usage</a></li>
			<li><a href="/cl_reference/">Command Line Reference</a></li>
			<li><a href="/writing_pipelines_modules/">Writing Pipelines &amp; Modules</a></li>
			<li><a href="/troubleshooting/">Troubleshooting</a></li>
		</ul>
	</nav>
	<section class="credits">
		<p>Viewing v0.3 of the docs. <a href="http://clusterflow.io/docs_versions">Other versions</a>.</p>
		<p>Cluster Flow was written by <a href="http://phil.ewels.co.uk" target="_blank">Phil Ewels</a> whilst working at the <a href="http://www.babraham.ac.uk" target="_blank">Babraham Institute</a>. He now maintains it from <a href="http://www.scilifelab.se" target="_blank">SciLifeLab</a> in Stockholm, Sweden.</p>
		<p class="logos">
			<a href="http://www.babraham.ac.uk" target="_blank"><img src="/resources/img/Babraham_logo.png" title="Babraham Institute"></a>
			<a href="http://www.scilifelab.se" target="_blank"><img src="/resources/img/SciLifeLab_logo.png" title="SciLifeLab"></a>
		</p>
		<p>This documentation is written using <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>.</p>
	</section>
</header>


<div class="mainpage mainpage-toc">

	<h1 class="page-title">Introduction</h1>

	

	<h1 id="what_is_cluster_flow">What is Cluster Flow?</h1>

<p>Cluster Flow is simple package to run pipelines in a cluster environment. It is comprised of several layers:</p>

<ul>
<li><code>cf</code>
<ul>
<li>The main cluster flow command. This is called to initiate a new pipeline run.</li>
</ul>
</li>

<li>Pipelines
<ul>
<li>Protocols that describe a series of modules to be run, along with any parameters</li>
</ul>
</li>

<li>Modules
<ul>
<li>The instructions for an individual task. These can be written in any language but must conform to a common API, described within this document.</li>
</ul>
</li>

<li>Runs
<ul>
<li>Created from the pipeline template for each file. Specifies configuration variables and traces output filenames.</li>
</ul>
</li>
</ul>

<p>Cluster Flow will set off multiple queued jobs on the cluster with queue dependencies as defined in the pipeline.</p>

<h1 id="how_does_cluster_flow_work">How does Cluster Flow work?</h1>

<p>A typical Cluster Flow run will work as follows:</p>

<ul>
<li>Pipelines and modules are written for Cluster Flow</li>

<li>The <code>cf</code> command is run, initiating a pipeline with a set of files</li>

<li>CF decides on a number of runs, each with a set of input files or URLs</li>

<li>Input filenames are checked to make sure that they all have the same file extension</li>

<li>If input files are fastq, CF tries to work out whether they’re paired end or single end from the filenames. It dies with an error if it finds a mixture</li>

<li>CF checks that the files exist (unless they’re URLs) and parses the pipeline file</li>

<li>A run file is created for each run. This contains the configuration variables, a copy of the pipeline used and the starting filenames with the associated id <code>start_000</code></li>

<li>CF submits all of the jobs to the cluster
<ul>
<li>Each module in each run has its own cluster job</li>

<li>Jobs are submitted with dependencies so that they execute in the order specified by the pipeline</li>
</ul>
</li>

<li>Main CF program finishes and exits</li>

<li>Modules execute, using the run file and the previous job ID to find their input files</li>

<li>STDERR is appended to the log files.
<ul>
<li>Commands should print their commands with <code>###CFCMD</code> prepended</li>

<li>Important messages should start with <code>###CF</code></li>
</ul>
</li>

<li>Upon the completion of each module, the filenames of any resulting output files are appended to the run file, along with the job ID so that the next module can find them</li>

<li>Once all pipeline modules have finished, core cluster flow notification modules run
<ul>
<li>These clean up the output file and send e-mails</li>
</ul>
</li>
</ul>

<h1 id="working_example">Working Example</h1>

<h2 id="find_whats_available">Find what’s available</h2>

<p>Often the first step when running a Cluster Flow pipeline is to remind yourself what is possible. Cluster Flow has a handful of functions to help you do this.</p>

<h3 id="available_pipelines">Available Pipelines</h3>

<p>Running</p>

<pre><code>cf --list_pipelines</code></pre>

<p>Will give a list of the pipelines installed in your copy of Cluster Flow, for example:</p>

<pre><code>================================
Cluster Flow - available pipelines
================================
Installed pipelines:
    Directory /clusterflow/0.1/pipelines/
	- bismark
	- bismark_pbat
	- bismark_singlecell
	- fastq_bismark
	- fastq_bowtie
	... (trimmed) ...</code></pre>

<h3 id="available_modules">Available Modules</h3>

<p>Cluster Flow can run single modules as well as strung-together in pipelines. The analgous command is:</p>

<pre><code>cf --list_modules</code></pre>

<p>Will give:</p>

<pre><code>================================
Cluster Flow - available modules
================================
Available modules:
    Directory /clusterflow/0.1/modules/
	- bismark_align
	- bismark_deduplicate
	- bismark_messy
	- bismark_methXtract
	- bismark_tidy
	- bowtie
	... (trimmed) ...</code></pre>

<h3 id="available_genomes">Available Genomes</h3>

<p>Most modules and pipelines require genomes. To see what is set up, run</p>

<pre><code>cf --list_genomes</code></pre>

<p>Will give:</p>

<pre><code>
================================
Cluster Flow - available genomes
================================

--------------------------------------------------
 /clusterflow/0.1/genomes.config
--------------------------------------------------

== Genome Paths ==
 Key                 Species             Assembly            Path
----------------------------------------------------------------------------------------------------
 GRCh37              Human               GRCh37         /Genomes/Human/GRCh37/
 GRCm38              Mouse               GRCm38         /Genomes/Mouse/GRCm38/
 NCBIM37             Mouse               NCBIM37        /Genomes/Mouse/NCBIM37/
 
== Bowtie Index Base Paths ==
 Key                 Species             Assembly            Path
----------------------------------------------------------------------------------------------------
 GRCh37              Human               GRCh37         /Genomes/Human/GRCh37/Homo_sapiens.GRCh37
 GRCm38              Mouse               GRCm38         /Genomes/Mouse/GRCm38/Mus_musculus.GRCm38
 NCBIM37             Mouse               NCBIM37        /Genomes/Mouse/NCBIM37/Mus_musculus.NCBIM37
 
== GTF File Paths ==
 Key                 Species             Assembly            Path
----------------------------------------------------------------------------------------------------
 GRCh37              Human               GRCh37         /Genomes/Human/GRCh37/Homo_sapiens.GRCh37.61.gtf
 GRCm38              Mouse               GRCm38         /Genomes/Mouse/GRCm38/Mus_musculus.GRCm38.70.gtf
 NCBIM37             Mouse               NCBIM37        /Genomes/Mouse/NCBIM37/Mus_musculus.NCBIM37.cleaned.gtf
</code></pre>

<h2 id="running_cluster_flow">Running Cluster Flow</h2>

<p>Once you’re happy with what’s available and what you want to run, you use the <code>cf</code> command.</p>

<p>For example, to run the <code>fastq_bowtie</code> pipeline (runs FastQC, trims the reads with Trim Galore! and then aligns with either bowtie 1 or bowtie 2, depending on the read length) on a selection of FastQ files, I would run:</p>

<pre><code>cf --genome GRCh37 fastq_bowtie *.fq.gz</code></pre>

<p>A typical output might be:</p>

<pre><code>Processing 3 files in 12 runs. Submitting 1 file per run.

Pipeline to be used:

#fastqc
#trim_galore
	#bowtie

Processing file 1 - filename_1.fq.gz
Processing file 2 - filename_2.fq.gz
Processing file 3 - filename_3.fq.gz
Your job 191764 (&quot;cf_sra_bowtie_1401371996_sra_fqdump_579&quot;) has been submitted
Your job 191765 (&quot;cf_sra_bowtie_1401371996_fastqc_975&quot;) has been submitted
Your job 191766 (&quot;cf_sra_bowtie_1401371996_trim_galore_490&quot;) has been submitted
Your job 191767 (&quot;cf_sra_bowtie_1401371996_bowtie_576&quot;) has been submitted
Your job 191768 (&quot;cf_sra_bowtie_1401371996_email_run_complete_840&quot;) has been submitted
Your job 191769 (&quot;cf_sra_bowtie_1401371996_sra_fqdump_816&quot;) has been submitted
Your job 191770 (&quot;cf_sra_bowtie_1401371996_fastqc_484&quot;) has been submitted
Your job 191771 (&quot;cf_sra_bowtie_1401371996_trim_galore_930&quot;) has been submitted
Your job 191772 (&quot;cf_sra_bowtie_1401371996_bowtie_281&quot;) has been submitted
Your job 191773 (&quot;cf_sra_bowtie_1401371996_email_run_complete_911&quot;) has been submitted
Your job 191774 (&quot;cf_sra_bowtie_1401371996_sra_fqdump_940&quot;) has been submitted
Your job 191775 (&quot;cf_sra_bowtie_1401371996_fastqc_805&quot;) has been submitted
Your job 191776 (&quot;cf_sra_bowtie_1401371996_trim_galore_634&quot;) has been submitted
Your job 191777 (&quot;cf_sra_bowtie_1401371996_bowtie_109&quot;) has been submitted
Your job 191778 (&quot;cf_sra_bowtie_1401371996_email_run_complete_922&quot;) has been submitted
Your job 191779 (&quot;cf_sra_bowtie_1401371996_email_pipeline_complete_127&quot;) has been submitted</code></pre>

<h2 id="checking_the_status_of_a_run">Checking the status of a run</h2>

<p>I can then check on the progress of my running job using:</p>

<pre><code>cf --qstat</code></pre>

<p>This will give the following report (much easier to read than the default <code>qstat</code>)</p>

<pre><code>======================================================================
 Cluster Flow Pipeline: sra_bowtie                                  
 Submitted:             30 seconds ago                              
 Working Directory:     /bi/group/bioinf/clusterflow/dev/test       
 ID:                    sra_bowtie_1401371972                       
======================================================================

 -  sra_fqdump                         [1 core] running for 16s
      - fastqc 
      - trim_galore 
           - bowtie 
                - email_run_complete 

 -  sra_fqdump                         [1 core]  [priority -500] queued for 27s
      - trim_galore 
           - bowtie 
                - email_run_complete 
                     - email_pipeline_complete 
      - fastqc 

 -  sra_fqdump                         [1 core]  [priority -500] queued for 27s
      - fastqc 
      - trim_galore 
           - bowtie 
                - email_run_complete</code></pre>

<h2 id="cancelling_a_pipeline">Cancelling a pipeline</h2>

<p>I might later realise that I messed something up with the input files, but since then I have started loads of other jobs. I don’t want to pull out all of the job IDs myself manually, so instead I can use Cluster Flow to delete the jobs belonging to the ID printed with the command above:</p>

<pre><code>cf --qdel sra_bowtie_1401371972</code></pre>

<p>Which might give the following outout:</p>

<pre><code>Deleting jobs from Pipeline id sra_bowtie_1401371972

16 jobs deleted:
ewelsp has registered the job 191748 for deletion
ewelsp has deleted job 191753
ewelsp has deleted job 191758
ewelsp has deleted job 191749
ewelsp has deleted job 191750
ewelsp has deleted job 191751
ewelsp has deleted job 191752
ewelsp has deleted job 191754
ewelsp has deleted job 191755
ewelsp has deleted job 191756
ewelsp has deleted job 191757
ewelsp has deleted job 191759
ewelsp has deleted job 191760
ewelsp has deleted job 191761
ewelsp has deleted job 191762
ewelsp has deleted job 191763</code></pre>

</div>

<aside class="sidebar">
	<nav id="subnav">
		<h3>Table of Contents</h3>
		<div class="toc"></div>
	</nav>
</aside>

	<footer>
		<section class="credits">
			<p>Currently viewing v0.3 of the docs. <a href="http://clusterflow.io/docs_versions">Find other versions</a>.</p>
			<p>Cluster Flow was written by <a href="http://phil.ewels.co.uk" target="_blank">Phil Ewels</a> whilst working at the <a href="http://www.babraham.ac.uk" target="_blank">Babraham Institute</a>. He now maintains it from <a href="http://www.scilifelab.se" target="_blank">SciLifeLab</a> in Stockholm, Sweden.</p>
			<p class="logos">
				<a href="http://www.babraham.ac.uk" target="_blank"><img src="/resources/img/Babraham_logo.png" title="Babraham Institute"></a>
				<a href="http://www.scilifelab.se" target="_blank"><img src="/resources/img/SciLifeLab_logo.png" title="SciLifeLab"></a>
			</p>
			<p>This documentation is written using <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>.</p>
		</section>
	</footer>
	<!-- jQuery & Boostrap -->
	<script src="/resources/jquery-1.11.1.min.js"></script>
	<script src="/resources/bootstrap/js/bootstrap.min.js"></script>
	<!-- Auto TOC Generator -->
	<script src="/resources/jekyll-table-of-contents-master/toc.js"></script>
	<script type="text/javascript">
		$(document).ready(function() {
		    $('.toc').toc({
				title: '',
				listType: 'ul'
			});
			$('.demo_gif').click(function(){
				if($(this).hasClass('inactive')){
					$(this).removeClass('inactive');
					$(this).addClass('active');
				} else {
					$(this).removeClass('active');
					$(this).addClass('inactive');
				}
			});
		});
	</script>
	<!-- Google Analytics-->
	<script>
	  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o), m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m) })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
	  ga('create', 'UA-51481908-1', 'ewels.github.io');
	  ga('send', 'pageview');
	</script>
</body>
</html>

