<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title> Cluster Flow |  Cluster Flow</title>
	<meta name="description" content="A pipelining tool to automate and standardise bioinformatics analyses on cluster environments"/>
	<meta name="author" content="Phil Ewels"/>
	<link rel="shortcut icon" href="/favicon.ico">

	<!-- Bootstrap and Google Fonts -->
	<link href="/resources/bootstrap/css/bootstrap.min.css" rel="stylesheet">
	<link href='http://fonts.googleapis.com/css?family=Lato:400,700,400italic' rel='stylesheet' type='text/css'>
	<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
	<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
	<!--[if lt IE 9]>
		<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
		<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
	<![endif]-->

	<!-- Site Stylesheet -->
	<link rel="stylesheet" href="/style.css">
</head>
<body>

<header>
	<section class="logo">
		<h1><a href="/"><img src="/resources/img/CF_logo.png" title="Cluster Flow"></a></h1>
		<p>A pipelining tool to automate and standardise bioinformatics analyses on cluster environments.</p>
	</section>
	<section class="download">
		<p>
			<a class="btn btn-default" href="/" title="Download Cluster Flow"><i class="glyphicon glyphicon-folder-open"></i> &nbsp; &nbsp; Download Cluster Flow</a>
			<a class="text-link" href="/" title="View on GitHub"><img src="/resources/img/GitHub-Mark/PNG/GitHub-Mark-Light-16px.png" title="View on GitHub"> &nbsp; View on GitHub</a>
		</p>
	</section>
	<nav id="nav">
		<ul>
			<li><a href="/introduction/">Introduction</a></li>
			<li><a href="/installation/">Installation Instructions</a></li>
			<li><a href="/usage/">General Usage</a></li>
			<li><a href="/cl_reference/">Command Line Reference</a></li>
			<li><a href="/writing_pipelines_modules/">Writing Pipelines &amp; Modules</a></li>
			<li><a href="/troubleshooting/">Troubleshooting</a></li>
		</ul>
	</nav>
	<section class="credits">
		<p>Viewing v0.3 of the docs. <a href="http://clusterflow.io/docs_versions">Other versions</a>.</p>
		<p>Cluster Flow was written by <a href="http://phil.ewels.co.uk" target="_blank">Phil Ewels</a> whilst working at the <a href="http://www.babraham.ac.uk" target="_blank">Babraham Institute</a>. He now maintains it from <a href="http://www.scilifelab.se" target="_blank">SciLifeLab</a> in Stockholm, Sweden.</p>
		<p class="logos">
			<a href="http://www.babraham.ac.uk" target="_blank"><img src="/resources/img/Babraham_logo.png" title="Babraham Institute"></a>
			<a href="http://www.scilifelab.se" target="_blank"><img src="/resources/img/SciLifeLab_logo.png" title="SciLifeLab"></a>
		</p>
		<p>This documentation is written using <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>.</p>
	</section>
</header>


<div class="mainpage">

	<h1 class="page-title">Cluster Flow</h1>

	

	<p class="lead">A command-line pipeline tool which uses common cluster managers to run bioinformatics analysis pipelines.</p>

<p>Benefits of using Cluster Flow:</p>

<ul>
<li>Routine analyses are very quick to run, for example: <code>cf --genome GRCh37 fastq_bowtie *fq.gz</code></li>

<li>Pipelines use identical parameters, standardising analysis and making results more reproducable</li>

<li>Integrated parallelisation tools help prevent your cluster becoming overloaded</li>

<li>All commands and output is logged in files for future reference</li>

<li>Intuitive commands and a comprehensive manual make Cluster Flow easy to use</li>

<li>Job monitoring tools and E-mail notifications allow you to keep track of your jobs</li>
</ul>

<p>How Cluster Flow differs from other pipeline tools:</p>

<ul>
<li>Very lightweight and flexible</li>

<li>Pipelines and configurations can easily be generated on a project-specific basis if required</li>

<li>New modules and pipelines are super easy to write (see video tutorial)</li>
</ul>

<p>Cluster Flow currently supports GRIDEngine (SGE), LSF and SLURM, plus it should be fairly easy to port to others.</p>
<div class='demo_gif inactive'>
<img src='resources/CF_mini.gif' title='mini CF demo' />
</div>
<h2 id="tutorial_videos">Tutorial Videos</h2>

<p>There are two tutorial videos up on YouTube, see below:</p>

<ul>
<li><a href="http://youtu.be/b2g_zQiz9ys">Usage / Installation Tutorial</a> - How to configure and run Cluster Flow</li>

<li><a href="http://youtu.be/aBHOcsA2M6w">Advanced Tutorial</a> - How to write your own pipelines and modules</li>
</ul>

<p>Please see the sidebar for access to the full Cluster Flow documentation. If you’re looking for a quick overview of what Cluster Flow is and how it works, see the <a href="/introduction/">Introduction</a>.</p>

<h2 id="contributing_to_cluster_flow">Contributing to Cluster Flow</h2>

<p>If you write a module or pipeline which could be of use to others, or modify the core Cluster Flow code in a helpful way, it would be great to merge those changes back into the core Cluster Flow project.</p>

<p>The easiest way to do this is to <a href="https://help.github.com/articles/fork-a-repo">fork the Cluster Flow repository</a>, make your changes, committing them and pushing them as you go. When you’ve finished, submit a <a href="https://help.github.com/articles/using-pull-requests">pull request</a> and the new code can be merged into the central project.</p>

<h2 id="change_log">Change Log</h2>
<h4 class='version'><a href='https://github.com/ewels/clusterflow/releases/tag/v0.3'><span class='label label-success version'>Version 0.3</span></a> <small>2014-07-11</small></h4>
<ul>
<li>New Stuff
<ul>
<li>Awesome new HTML report e-mails
<ul>
<li>Much more readable HTML report e-mails which look super-snazzy (see <a href="http://ewels.github.io/clusterflow/example_report_good.html">example</a>)</li>

<li>Any errors are highlighted making them quick to identify (see <a href="http://ewels.github.io/clusterflow/example_report_bad.html">example</a>)</li>

<li>Custom strings set in the config can flagged as <a href="http://ewels.github.io/clusterflow/example_report_highlights.html">highlights</a> or as <a href="http://ewels.github.io/clusterflow/example_report_warnings.html">warnings</a></li>

<li>Designed to work on desktop and mobile phone screens</li>
</ul>
</li>

<li>Cluster Flow now re-orders the log file so that output from different modules doesn’t overlap
<ul>
<li>Made each module prepend its stdout and sterr with a CF module flag</li>

<li>Made the <code>cf_run_finished</code> module parse the above flag and print out module by module</li>
</ul>
</li>

<li>Rewrote how the environment module loading works - now much more robust
<ul>
<li>Uses <code>modulecmd</code> to prepare perl syntax commands</li>

<li>Moved code into the <code>load_environment_modules</code> function in Helpers.pm</li>
</ul>
</li>

<li>Added environment module aliases
<ul>
<li>This allows you to load specific environment module versions or use different names to those specified within CF modules</li>

<li>eg. Replace <code>fastqc</code> with <code>FastQC/0.11.2</code></li>
</ul>
</li>
</ul>
</li>

<li>Updates
<ul>
<li>Made the <code>bismark_methXtract</code> module create genome-wide coverage reports if GTF path is available</li>

<li>Added the <code>-q</code> parameter to the FastQC module to make the log files cleaner</li>

<li>Removed the now uneccesary <code>bismark_tidy</code> module and renamed <code>bismark_messy</code> to <code>bismark_report</code></li>
</ul>
</li>

<li>Bugs Squashed
<ul>
<li>Fixed dependency bug introduced in v0.2 which was making all downloads fire simultaneously</li>

<li>Fixed issue where modules using the CF::Constants Perl Module couldn’t load the central config file</li>

<li>Fixed typo in environment module loading in <code>bismark_align</code> module</li>

<li>Reordered loading of the environment modules in <code>trim_galore</code> so that FastQC is loaded first, fixing dependency issues</li>
</ul>
</li>
</ul>
<h4 class='version'><a href='https://github.com/ewels/clusterflow/releases/tag/v0.2'><span class='label label-success version'>Version 0.2</span></a> <small>2014-05-29</small></h4>
<ul>
<li>New Stuff
<ul>
<li>Now compatable with SLURM</li>

<li>Customise batch job commands in the config (see <a href="/installation/#making_cluster_flow_work_with_your_environment">docs</a> for more info)</li>

<li>Created new GitHub pages website to hold documentation: <a href="http://ewels.github.io/clusterflow/">http://ewels.github.io/clusterflow/</a></li>
</ul>
</li>

<li>Updates
<ul>
<li>Ported repository to github: <a href="https://github.com/ewels/clusterflow">https://github.com/ewels/clusterflow</a></li>

<li>Wrote new readme for github</li>
</ul>
</li>

<li>Bugs Squashed
<ul>
<li>Custom modules in <code>~/clusterflow/modules/</code> weren’t being found</li>

<li>General code clean-ups all over the place</li>
</ul>
</li>
</ul>
<h4 class='version'><a href='https://github.com/ewels/clusterflow/releases/tag/v0.1'><span class='label label-success version'>Version 0.1</span></a> <small>2014-04-25</small></h4>
<ul class="changelog">
<li>The first public release of Cluster Flow, although it’s been in use at the Babraham Institute for around 6 months. It’s been in heavy development throughout that time and is now approaching a state of being relatively stable.</li>
</ul>

</div>

	<footer>
		<section class="credits">
			<p>Currently viewing v0.3 of the docs. <a href="http://clusterflow.io/docs_versions">Find other versions</a>.</p>
			<p>Cluster Flow was written by <a href="http://phil.ewels.co.uk" target="_blank">Phil Ewels</a> whilst working at the <a href="http://www.babraham.ac.uk" target="_blank">Babraham Institute</a>. He now maintains it from <a href="http://www.scilifelab.se" target="_blank">SciLifeLab</a> in Stockholm, Sweden.</p>
			<p class="logos">
				<a href="http://www.babraham.ac.uk" target="_blank"><img src="/resources/img/Babraham_logo.png" title="Babraham Institute"></a>
				<a href="http://www.scilifelab.se" target="_blank"><img src="/resources/img/SciLifeLab_logo.png" title="SciLifeLab"></a>
			</p>
			<p>This documentation is written using <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>.</p>
		</section>
	</footer>
	<!-- jQuery & Boostrap -->
	<script src="/resources/jquery-1.11.1.min.js"></script>
	<script src="/resources/bootstrap/js/bootstrap.min.js"></script>
	<!-- Auto TOC Generator -->
	<script src="/resources/jekyll-table-of-contents-master/toc.js"></script>
	<script type="text/javascript">
		$(document).ready(function() {
		    $('.toc').toc({
				title: '',
				listType: 'ul'
			});
			$('.demo_gif').click(function(){
				if($(this).hasClass('inactive')){
					$(this).removeClass('inactive');
					$(this).addClass('active');
				} else {
					$(this).removeClass('active');
					$(this).addClass('inactive');
				}
			});
		});
	</script>
	<!-- Google Analytics-->
	<script>
	  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o), m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m) })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
	  ga('create', 'UA-51481908-1', 'ewels.github.io');
	  ga('send', 'pageview');
	</script>
</body>
</html>

