<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="utf-8">
	<meta http-equiv="X-UA-Compatible" content="IE=edge">
	<meta name="viewport" content="width=device-width, initial-scale=1">
	<title> Installation Instructions |  Cluster Flow</title>
	<meta name="description" content="A pipelining tool to automate and standardise bioinformatics analyses on cluster environments"/>
	<meta name="author" content="Phil Ewels"/>
	<link rel="shortcut icon" href="/favicon.ico">

	<!-- Bootstrap and Google Fonts -->
	<link href="/resources/bootstrap/css/bootstrap.min.css" rel="stylesheet">
	<link href='http://fonts.googleapis.com/css?family=Lato:400,700,400italic' rel='stylesheet' type='text/css'>
	<!-- HTML5 Shim and Respond.js IE8 support of HTML5 elements and media queries -->
	<!-- WARNING: Respond.js doesn't work if you view the page via file:// -->
	<!--[if lt IE 9]>
		<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
		<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
	<![endif]-->

	<!-- Site Stylesheet -->
	<link rel="stylesheet" href="/style.css">
</head>
<body>

<header>
	<section class="logo">
		<h1><a href="/"><img src="/resources/img/CF_logo.png" title="Cluster Flow"></a></h1>
		<p>A pipelining tool to automate and standardise bioinformatics analyses on cluster environments.</p>
	</section>
	<section class="download">
		<p>
			<a class="btn btn-default" href="/" title="Download Cluster Flow"><i class="glyphicon glyphicon-folder-open"></i> &nbsp; &nbsp; Download Cluster Flow</a>
			<a class="text-link" href="/" title="View on GitHub"><img src="/resources/img/GitHub-Mark/PNG/GitHub-Mark-Light-16px.png" title="View on GitHub"> &nbsp; View on GitHub</a>
		</p>
	</section>
	<nav id="nav">
		<ul>
			<li><a href="/introduction/">Introduction</a></li>
			<li><a href="/installation/">Installation Instructions</a></li>
			<li><a href="/usage/">General Usage</a></li>
			<li><a href="/cl_reference/">Command Line Reference</a></li>
			<li><a href="/writing_pipelines_modules/">Writing Pipelines &amp; Modules</a></li>
			<li><a href="/troubleshooting/">Troubleshooting</a></li>
		</ul>
	</nav>
	<section class="credits">
		<p>Viewing v0.3 of the docs. <a href="http://clusterflow.io/docs_versions">Other versions</a>.</p>
		<p>Cluster Flow was written by <a href="http://phil.ewels.co.uk" target="_blank">Phil Ewels</a> whilst working at the <a href="http://www.babraham.ac.uk" target="_blank">Babraham Institute</a>. He now maintains it from <a href="http://www.scilifelab.se" target="_blank">SciLifeLab</a> in Stockholm, Sweden.</p>
		<p class="logos">
			<a href="http://www.babraham.ac.uk" target="_blank"><img src="/resources/img/Babraham_logo.png" title="Babraham Institute"></a>
			<a href="http://www.scilifelab.se" target="_blank"><img src="/resources/img/SciLifeLab_logo.png" title="SciLifeLab"></a>
		</p>
		<p>This documentation is written using <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>.</p>
	</section>
</header>


<div class="mainpage mainpage-toc">

	<h1 class="page-title">Installation Instructions</h1>

	

	<h1 id="installation_overview">Installation Overview</h1>

<p>The heart of Cluster Flow is a simple Perl script. To get it up and running on your system there a few simple steps:</p>

<ul>
<li>Download the code!</li>

<li>Make the main script accessible by adding it to your <code>PATH</code> with environment modules or other methods <em>(optional)</em></li>

<li>Set up the configuration file to work with your cluster environment</li>

<li>Add some genome paths</li>
</ul>

<h2 id="downloading_cluster_flow">Downloading Cluster Flow</h2>

<p>To get started with Cluster Flow, first download the source files using the links on the left. Stable releases are tagged as downloads. The main github branch will typically contain the latest stable release, with development work going on within development branches.</p>

<h2 id="configuring_your_">Configuring your <code>PATH</code></h2>

<p>Linux systems use an environment variable called <code>PATH</code> to keep the locations of executable binaries. If you add the Cluster Flow directory to this, you’ll be able to call the Cluster Flow <code>cf</code> file from anywhere.</p>

<h3 id="environment_modules">Environment modules</h3>

<p>Many compute clusters use <a href="http://modules.sourceforge.net/">environment modules</a> to manage the user’s <code>PATH</code>.</p>

<p>If your system uses environment modules, you can create a new module file to add Cluster Flow to the list of available modules. A typical module file example is given below - note that you’ll need to change the file paths to where the Cluster Flow files are kept. The code below takes the version of Cluster Flow from the module script filename.</p>

<pre><code>#%Module1.0#####################################################################
##
## clusterflow modulefile
##

source $env(MODULE_INCLUDE)/functions.tcl
getCluster

set components [ file split [ module-info name ] ]
set version [ lindex $components 1 ]

set     modroot          /path/to/clusterflow/$version

proc ModulesHelp { } {
        global version modroot

        puts stderr &quot;\tclusterflow - use clusterflow $version&quot;
        puts stderr &quot;\n\tVersion $version\n&quot;
}

module-whatis   &quot;Loads clusterflow environment.&quot;

# Only one version at a time
conflict clusterflow

#Log loading to syslog
logToSyslog

if [module-info mode load] {
    prepend-path        PATH            $modroot
}

if [module-info mode remove] {
    remove-path         PATH            $modroot
}</code></pre>

<p>Once done, you may need to rebuild your module cache. You can load the new module by running</p>

<pre><code>module load clusterflow</code></pre>

<p>You may also wish to create a symlinked directory called cf so that <code>module load cf</code> also works.</p>

<h3 id="manually_adding_cluster_flow_to_the_">Manually adding Cluster Flow to the <code>PATH</code></h3>

<p>You can manually add the directory containing cluster flow to your <code>PATH</code> with the following command :</p>

<pre><code>PATH=/path/to/clusterflow:$PATH</code></pre>

<h3 id="checking_its_worked">Checking it’s worked</h3>

<p>For both of the above methods, you can check that the system can find Cluster Flow with the following command:</p>

<pre><code>which cf</code></pre>

<p>You should see the directory containing the Cluster Flow files.</p>

<h2 id="configuring_cluster_flow">Configuring Cluster Flow</h2>

<h3 id="cluster_flow_config_files">Cluster Flow config files</h3>

<p>Cluster flow will search three locations for a config file every time it is run. Variables found in each file can override those read from a previous config file. They are, in order of priority:</p>

<ul>
<li><code>&lt;working directory&gt;/clusterflow.config</code>
<ul>
<li>A config file found in the current working directory when a pipeline is executed has top priority, trumped only by command line parameters</li>
</ul>
</li>

<li><code>~/clusterflow.config</code>
<ul>
<li>A config file in your home directory can be used to set parameters such as notification level and e-mail address</li>
</ul>
</li>

<li><code>&lt;installation directory&gt;/clusterflow.config</code>
<ul>
<li>A config file in the Cluster Flow installation directory is ideal for common settings specific to the environment</li>
</ul>
</li>
</ul>

<p>Config files contain key: value pairs. Syntax is as follows: <code>@key value</code> (tab delimited, one per line). Cluster Flow ships with an example config file called <a href="https://github.com/ewels/clusterflow/blob/master/clusterflow.config.example" title="Browse the example config file on GitHub">clusterflow.config.example</a></p>

<p>Typically, there will be a config file in the installation directory which contains the settings that make Cluster Flow work, then each user will have a personal configuration file in their home directory containing settings such as a notification e-mail address.</p>

<h3 id="making_cluster_flow_work_with_your_environment">Making Cluster Flow work with your environment</h3>

<p>The key things to set up when installing Cluster Flow are the variables that dictate how CF should interact with your cluster - what commands it should use to submit jobs.</p>

<p>Cluster Flow currently supports GRIDEngine (SGE), SLURM and LSF. You can choose which one you’re running with the following:</p>

<pre><code>/* Options: GRIDEngine, SLURM or LSF */
@cluster_environment    SLURM</code></pre>

<p>In most cases, that should be enough to get Cluster Flow to work! However, some people have some specific variables that need to be submitted with batch jobs (eg. project identifiers, time limits, other custom flags). If this is the case, the job submission command can be customised with the <code>@custom_job_submit_command</code> config variable.</p>

<p>To use this, enter your typical submission command with the following placeholders which will be replaced at run time:</p>

<ul>
<li><code>{{command}}</code>
<ul>
<li>The actual command which will be run to execute the module file</li>
</ul>
</li>

<li><code>{{job_id}}</code>
<ul>
<li>The unique identifier which will be assigned to use job dependencies</li>
</ul>
</li>

<li><code>{{outfn}}</code>
<ul>
<li>The filename of the log file to capture <code>STDOUT</code></li>
</ul>
</li>

<li><code>{{cores}}</code>
<ul>
<li>How many cores to assign</li>
</ul>
</li>

<li><code>{{mem}}</code>
<ul>
<li>How much memory to assign</li>
</ul>
</li>

<li><code>{{priority}}</code>
<ul>
<li>A priority to set, defined in the config file</li>
</ul>
</li>

<li><code>{{email}}</code>
<ul>
<li>The user’s e-mail address for cluster job notifications (if set)</li>
</ul>
</li>

<li><code>{{notifications}}</code>
<ul>
<li>A string describing which notifications to be sent (syntax depends on environment set above)</li>
</ul>
</li>
</ul>

<p>For example:</p>

<pre><code>@custom_job_submit_command      sbatch  -A MY_PROJECT_ID -t 2-00:00:00 -p core -n {{cores}} --open-mode=append -o {{outfn}} -J {{job_id}} {{notifications}} --wrap=&quot;{{command}}&quot;</code></pre>

<p>Cluster Flow will generate it’s own sensible default if this isn’t set, so it’s worth trying it without first.</p>

<p><strong>Note</strong> - Cluster Flow will append the job dependency strings to the end of your custom command which are system specific, so it’s important that <code>@cluster_environment</code> is still correct.</p>

<h3 id="system_specific_gotchas">System specific Gotchas</h3>

<p>There are a few non-obvious things that have to be set correctly for some environments to work..</p>

<ul>
<li>GridEngine / SGE
<ul>
<li>GridEngine defaults to using <code>csh</code> to execute commands instead of the more standard <code>bash</code>. This will break some things. To override this behaviour, include the parameter <code>-S /bin/bash</code> for the commands</li>
</ul>
</li>
</ul>

<h2 id="adding_genome_paths">Adding genome paths</h2>

<p>Many modules within Cluster Flow require reference genomes for alignment. You can do this manually (probably quicker if you have lots to add) or you can use the interactive wizard that comes with Cluster Flow - just run:</p>

<pre><code> cf --add_genome</code></pre>

<p>Genome paths are stored within files called <code>genomes.config</code> which are stored in the same directories as <code>clusterflow.config</code>. Within this file, each path is described with <code>@genome_path</code> followed by a unique IDy used when submitting the Cluster Flow run (eg. <code>--genome GRCh37</code>). These are then followed by an absolute path. Optionally, species and assembly can be added after this (see <a href="https://github.com/ewels/clusterflow/blob/master/genomes.config.example" title="Browse the example genomes file on GitHub">example</a>).</p>

<p>There are four types of paths that can be specified:</p>

<ul>
<li><code>@genome_path</code>
<ul>
<li>The directory containing a reference genome</li>
</ul>
</li>

<li><code>@bowtie_path</code>
<ul>
<li>The file name base of a set of bowtie indices</li>
</ul>
</li>

<li><code>@bowtie2_path</code>
<ul>
<li>The file name base of a set of bowtie 2 indices</li>
</ul>
</li>

<li><code>@gtf_path</code>
<ul>
<li>The file name of a GTF file for a given genome.</li>
</ul>
</li>
</ul>

<p>All four types of path should share genome keys if applicable. The fields should be separated by a tab character. Cluster Flow ships with an example genomes file called <a href="https://github.com/ewels/clusterflow/blob/master/genomes.config.example" title="Browse the example genomes file on GitHub">genomes.config.example</a></p>

<h2 id="environment_module_aliases">Environment Module Aliases</h2>

<p>If using environment modules, you may get some errors claiming that certain tools are not installed. If you think that you do have that tool installed, it could be because of a minor difference in the module name (eg. <code>fastqc</code> versus <code>FastQC</code>). To avoid having to change the name of your modules, you can configure aliases in your configuration file. You can also use these aliases to specify specific software versions for Cluster Flow.</p>

<p>Aliases are added with the <code>@environment_module_alias</code> tag. For example:</p>

<pre><code>@environment_module_alias	fastqc	FastQC/0.11.2
@environment_module_alias	trim_galore	TrimGalore</code></pre>

<h2 id="running_cluster_flow">Running Cluster Flow</h2>

<p>Ok, you’re done! Cluster Flow should now work (see <a href="/usage/">General Usage</a> for instructions on how to run Cluster Flow). If you get any weird errors, have a look through the <a href="/troubleshooting/">Troubleshooting</a> section and if in doubt, drop me an e-mail. Good luck!</p>
<hr />
<h1 id="config_file_reference">Config File reference</h1>

<p>The following section describes the available variables that can be set in the config file. For an example, see the <a href="https://github.com/ewels/clusterflow/blob/master/clusterflow.config.example" title="Browse the example config file on GitHub">clusterflow.config.example</a> file that comes bundled with Cluster Flow.</p>

<h3 id="email">@email</h3>

<p>Sets your e-mail address, used for e-mail notifications.</p>

<h3 id="check_updates">@check_updates</h3>

<p>Cluster Flow can automatically check for new versions. If an update is available, it will print a notification each time you run a job. You can specify how often Cluster Flow should check for updates with this parameter. The syntax is a number followed by <code>d</code>, <code>w</code>, <code>m</code> or <code>y</code> for days, weeks, months or years. Cluster Flow will check for an update at runtime if this period or more has elapsed since you last ran it. You can disable update checks and alerts by setting <code>@check_updates 0</code> in your <code>~/clusterflow.config</code> file.</p>

<p>You can get Cluster Flow to manually check for updates by running cf <code>--check_updates</code></p>

<h3 id="split_files">@split_files</h3>

<p>The default number of input files to send to each run. Typically set to 1.</p>

<h3 id="priority">@priority</h3>

<p>The priority to give to cluster jobs.</p>

<h3 id="max_runs">@max_runs</h3>

<p>The maximum number of parallel runs that cluster flow will set off in one go. Default is 12 to avoid swamping the cluster for all other users.</p>

<h3 id="notification">@notification</h3>

<p>Multiple <code>@notification</code> key pairs can be set with the following values:</p>

<ul>
<li>complete
<ul>
<li>An e-mail notification is sent when all processing for all files has finished</li>
</ul>
</li>

<li>run
<ul>
<li>An e-mail is sent when each run finishes (each set of input files)</li>
</ul>
</li>

<li>end
<ul>
<li>A qsub notification e-mail is sent when each cluster job ends. Likely to result in a full inbox!</li>
</ul>
</li>

<li>suspend
<ul>
<li>A qsub notification e-mail is sent if a job is suspended</li>
</ul>
</li>

<li>abort
<ul>
<li>A qsub notification e-mail is sent if a job is aborted</li>
</ul>
</li>
</ul>

<p>Cluster Flow sends the <code>run</code> and <code>complete</code> notifications using the <code>cf_run_finished</code> and <code>cf_runs_all_finished</code> modules. These modules handle several tasks, such as cleaning useless warning messages from log files. E-mails contain the contents of all log files, plus a section at the top of highlighted messages, specified within log messages by being prefixed with <code>###CF</code>.</p>

<h3 id="total_cores">@total_cores</h3>

<p>The total number of cores available to a Cluster Flow pipeline. Modules are given a recommended number of cores so that resources can be allocated without swamping the cluster.</p>

<h3 id="total_mem">@total_mem</h3>

<p>The total amount of memory available to a Cluster Flow pipeline. Modules are given a recommended quota so that resources can be allocated without swamping the cluster.</p>

<h3 id="cluster_environment">@cluster_environment</h3>

<p>Cluster Flow can submit jobs to both GRIDEngine and LSF cluster management systems. This configuration variable sets which environment to use. The possible options are:</p>

<ul>
<li>GRIDEngine</li>

<li>SLURM</li>

<li>LSF</li>
</ul>

<p>It is worth noting that Cluster Flow has been developed on GRIDEngine and SLURM Clusters so is likely to be most robust on similar setups. The <code>--qstat</code> and <code>--qstatall</code> parameters do not currently work with LSR. If you’d like to have a go at implementing it, that would be great!</p>

<h3 id="ignore_modules">@ignore_modules</h3>

<p>If you do not use environment modules on your system, you can prevent Cluster Flow from trying to use them (and giving a warning) by adding this line to your config file.</p>

<h3 id="environment_module_alias">@environment_module_alias</h3>

<p>See above for instructions on how to use this tag.</p>

<h3 id="log_highlight_string_and_log_warning_string">@log_highlight_string and @log_warning_string</h3>

<p>To pull out specific highlights or warnings from log files, you can specify search strings with these tags. If found, the e-mail will be highlighted accordingly and the lines from the log file will be displayed at the top of the report e-mail.</p>

<p>For example:</p>

<pre><code>@log_highlight_string at least one reported alignment
@log_warning_string job failed</code></pre>

</div>

<aside class="sidebar">
	<nav id="subnav">
		<h3>Table of Contents</h3>
		<div class="toc"></div>
	</nav>
</aside>

	<footer>
		<section class="credits">
			<p>Currently viewing v0.3 of the docs. <a href="http://clusterflow.io/docs_versions">Find other versions</a>.</p>
			<p>Cluster Flow was written by <a href="http://phil.ewels.co.uk" target="_blank">Phil Ewels</a> whilst working at the <a href="http://www.babraham.ac.uk" target="_blank">Babraham Institute</a>. He now maintains it from <a href="http://www.scilifelab.se" target="_blank">SciLifeLab</a> in Stockholm, Sweden.</p>
			<p class="logos">
				<a href="http://www.babraham.ac.uk" target="_blank"><img src="/resources/img/Babraham_logo.png" title="Babraham Institute"></a>
				<a href="http://www.scilifelab.se" target="_blank"><img src="/resources/img/SciLifeLab_logo.png" title="SciLifeLab"></a>
			</p>
			<p>This documentation is written using <a href="http://jekyllrb.com/" target="_blank">Jekyll</a>.</p>
		</section>
	</footer>
	<!-- jQuery & Boostrap -->
	<script src="/resources/jquery-1.11.1.min.js"></script>
	<script src="/resources/bootstrap/js/bootstrap.min.js"></script>
	<!-- Auto TOC Generator -->
	<script src="/resources/jekyll-table-of-contents-master/toc.js"></script>
	<script type="text/javascript">
		$(document).ready(function() {
		    $('.toc').toc({
				title: '',
				listType: 'ul'
			});
			$('.demo_gif').click(function(){
				if($(this).hasClass('inactive')){
					$(this).removeClass('inactive');
					$(this).addClass('active');
				} else {
					$(this).removeClass('active');
					$(this).addClass('inactive');
				}
			});
		});
	</script>
	<!-- Google Analytics-->
	<script>
	  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){(i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o), m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m) })(window,document,'script','//www.google-analytics.com/analytics.js','ga');
	  ga('create', 'UA-51481908-1', 'ewels.github.io');
	  ga('send', 'pageview');
	</script>
</body>
</html>

